syntax = "proto3";

package epoch_folio;

import "common.proto";

option go_package = "github.com/epochlab/epoch_protos/go/epoch_folio";
option java_package = "com.epochlab.epoch_protos.epoch_folio";
option java_outer_classname = "TableDefProto";
option csharp_namespace = "EpochLab.EpochProtos.EpochFolio";

// Sub-category definition
message SubCategoryDef {
  EpochFolioCategory type = 1;
  string name = 2;
}

// Category definition with sub-categories
message CategoryDef {
  EpochFolioCategory type = 1;
  string name = 2;
  repeated SubCategoryDef sub_categories = 3;
}

// Column definition for tables
message ColumnDef {
  string id = 1;
  string name = 2;
  EpochFolioType type = 3;
}

// Arrow table data representation
message TableData {
  // Schema definition
  repeated ColumnDef schema = 1;
  
  // Row data - each row contains values in the same order as schema
  repeated TableRow rows = 2;
}

// Single row of table data
message TableRow {
  repeated Scalar values = 1;
}

// Table definition with data
message Table {
  EpochFolioDashboardWidget type = 1;
  EpochFolioCategory category = 2;
  string title = 3;
  repeated ColumnDef columns = 4;
  TableData data = 5;
}

// Card data for dashboard cards
message CardData {
  string title = 1;
  Scalar value = 2;
  EpochFolioType type = 3;
  uint64 group = 4;
}

// Card definition for dashboard
message CardDef {
  EpochFolioDashboardWidget type = 1;
  EpochFolioCategory category = 2;
  repeated CardData data = 3;
  uint64 group_size = 4;
}
